File: 0323-number-of-connected-components-in-an-undirected-graph/disjoint_set_soln.cpp
class Solution {
public:
    // Disjoint set union approach
    // Time complexity: O(E.alpha(n)) for iteration over E edges and O(alpha(n)) for Inverse Ackermann function.
    // Space complexity: O(V) for parent array 
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<int> parent(n), rank(n);
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            rank[i] = 1;
        }
        int numComps = n;
        
        for (auto e : edges) {
            numComps -= addEdge(e[0], e[1], parent, rank);
        }
        return numComps;
    }
private:
    int findPar(int x, vector<int>& parent) {
        if (x == parent[x]) {
            return x;
        }
        return parent[x] = findPar(parent[x], parent);
    }
    
    int addEdge(int x, int y, vector<int>& parent, vector<int>& rank) {
        x = findPar(x, parent);
        y = findPar(y, parent);
        
        if (x == y) {
            return 0;
        }
        else {
            if (rank[x] > rank[y]) {
                parent[y] = x;
                rank[x] += rank[y];
            }
            else {
                parent[x] = y;
                rank[y] += rank[x];
            }
            return 1;
        }
        
    }
};

File: 0323-number-of-connected-components-in-an-undirected-graph/disjoint_set_soln.py
# Time complexity: O(V) to build the parent and size arrays + O(E.alpha(V)) to perform combine operation for E edges = O(V + E.alpha(V))
# Space complexity: O(V) for parent and size arrays
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        parent = [_ for _ in range(n)]
        size = [1]*n
        num_comps = n

        def findPar(node):
            if node != parent[node]:
                parent[node] = findPar(parent[node])
            
            return parent[node]

        def combine(x, y):
            par_x = findPar(x)
            par_y = findPar(y)

            if par_x == par_y:
                return 0
            else:
                if size[par_x] > size[par_y]:
                    parent[par_y] = par_x
                    size[par_x] += size[par_y]
                else:
                    parent[par_x] = par_y
                    size[par_y] += size[par_x]
                return 1

        for e in edges:
            num_comps -= combine(e[0], e[1])

        return num_comps